### DISCUSSION TITLE
Add min_custom and max_custom methods to Array class

### DISCUSSION THREAD

**Discussion by jwien001** [godotengine/godot-proposals#6609] [Scripting]
A common problem is finding the element in an array of objects with the highest or lowest value. An example might be finding the enemy that is closest to the player. Currently, it is possible to implement this by sorting the array with `sort_custom` and taking the first element of the sorted array, but that is overkill as the array only needs to be iterated over once. You can also implement this manually with a couple of variables and a `for` loop, but that's inconvenient, especially when this functionality is needed in multiple places.

Adding methods `Variant min_custom(func: Callable)` and `Variant max_custom(func: Callable)` to Array that returned the Variant in the array with the minimum/maximum value as returned by the given Callable would be more performant than sorting in cases where only the min/max element is needed and easier for developers to use than writing their own looping code.

**Comment by AThousandShips** [comment]
You can achieve this by using `reduce`

**Reply by AThousandShips** [reply]
Like so (assuming you want the *first* smallest, as `min` does):
```gdscript
arr.reduce(func (a, b): return a if not cmp(b, a) else b)
```

If the specific minimum doesn't matter you can do the simpler:
```gdscript
arr.reduce(func (a, b): return a if cmp(a, b) else b)
```

Just negate if you want the max, or swap the arguments

**Reply by Calinou** [reply]
@AThousandShips This should be documented in the `min()` and `max()` class reference :slightly_smiling_face:

**Reply by AThousandShips** [reply]
Then I'll write up an example for this!

**Reply by AThousandShips** [reply]
Open godotengine/godot#75625

**Reply by ra314** [reply]
I found this discussion from #7961. It would be nice to still have these dedicated functions even if the same functionality can be achieved by using reduce. The dedicated functions can just be aliases for the equivalent reduce function. Syntactic sugar is a great thing, unless it gets in the way of performance I don't see why it should be excluded.
